hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java

```
@@ -871,6 +877,15 @@ public Configuration(Configuration other) {
     setQuietMode(other.getQuietMode());
   }
 
+  protected synchronized void setPropListeners(
+      BiConsumer<String, String> propAddListener,
+      Consumer<String> propRemoveListener
+  ) {
+    this.properties = null;
```

Here, you could accept null values for the consumers, at least there's no prevention for them to be null.
In getProps, the PropWithListener is created if any of the listeners are not null (so the other one is allowed to be null).
Calling setProperty on PropWithListener is not checking if those fields are null, which is dangerous.
Either prevent them to be null in the constructor (fail fast) or do a null check in setProperty.

Could you please also add a testcase to cover the null scenarios?

---

I don't think it's okay to use the assert statement here, see: https://www.baeldung.com/java-avoid-null-check#assertions
Objects.requireNonNull is fine I think.